doc/decisions/change_tracking.md

```diff
@@ -0,0 +1,149 @@
+   change tracking.
+
+2. Fix the change tracking in the plugins by remembering the parent key for each operation.
+
+   This fixes the consistency problem and is transparent for developers and users.
+   However, the duplicated data and maintenance problem is not solved by this approach.
+
+3. Do the per-parent-key tracking within `libelektra-kdb`, within the `kdbGet` and `kdbSet` operations.
+
+   This is the first option that does not break any of the constraints.
+
+   Essentially, we'd have a 'giant' hashmap in the form of (parent key)->(KeySet) for all parent keys
+   that were used for a `kdbGet` operation in the lifetime of the `kdb` instance. We also need to
+   update it on `kdbSet` so that a future `kdbSet` operation without a `kdbGet` will also work.
+
+   When change tracking is enabled, this one will have the most memory overhead, as we need to deep-dup  
+   every key we have read or stored.
+
+4. Do the per-parent-key tracking within `libelektra-kdb`, but with meta-keys.
+
+   Essentially the same approach as above, but instead of deep-duping, we add the original value
+   as a meta-key to every key. Not yet clear how we handle changes to meta-data then.
+
+5. Outsource the change tracking into a separate plugin.
+
+   Essentially the same as (3), just that it is implemented within a plugin and not
+   `libelektra-kdb`. This will be a hook plugin, and will be called within `kdbGet` and `kdbSet` accordingly.
+   It will also need to export a hook-method to get the changeset.
+
+6. Do copy-on-write change tracking within `libelektra-core`.
+
+   This option will also not break any of the constraints.
+
+   The idea here is that we extend the `KeySet` and `Key` structs with additional fields.
+   For `KeySet` we need to track the following info:
+
+   - What keys have been removed
+   - What keys have been added
+   - Whether tracking is enabled for this KeySet
+
+   For `Key` we need to track:
+
+   - Original value of the key
+   - Size of the original value (for binary keys)
+   - Whether tracking is enabled for this key
+
+   The tracking itself would be done within the `ks*` and `key*` methods, after checking if it is enabled.
+   It would also transparently work for meta-data, as meta-data itself is implemented as a keyset with keys.
+
+   Two downsides of this method have been identified by @kodebach:
+
+   - It adds functionality to `libelektra-core` which is supposed to be minimal
+   - Adding fields to the structs causes a slight memory overhead, even with tracking turned off.
+     While neglectable for `KeySet` due to the low amount of keysets in typical applications,
+     it may be noticable for `Key`. On a 64-bit system we'd add 8+8=16 bytes to it.
+     To put this in perspective, the current size of the `Key` struct is 64 bytes,
+     so we'd add 25% overhead to an empty key. However, this percentage will be much lower in
+     a real-world application, as the usefulness of an empty key is very low.
```

Comment:
@atmaxinger can you please elaborate which 16 bytes you mean? Isn't a single pointer to "change-tracking data" enough? And maybe it even can be in meta-data, not needing any extra bytes in _Key.

Reply:
8 bytes for the pointer, 8 bytes for the size.

I won't comment on the rest, because this clearly goes into too much detail. If the problem is now clear, this PR should be merged and this should be addressed in the next PR.